R/auth.R

```diff
@@ -0,0 +1,292 @@
+#' 
+#' @description 
+#' There are three ways that you can authenticate with the twitter API:
+#' 
+#' * `rtweet_user()` interactively authenticates an existing twitter user. 
+#'   This form is most appropriate if you want rtweet to control your
+#'   twitter account.
+#'   
+#' * `rtweet_app()` authenticates as a twitter application. An application can't 
+#'    perform actions (i.e. it can't tweet) but otherwise has generally higher 
+#'    rate limits (i.e. you can do more searches). See details
+#'    at <https://developer.twitter.com/en/docs/basics/rate-limits.html>.
+#'    This form is most appropriate if you are collecting data. 
+#'    
+#' * `rtweet_bot()` authenticates as bot that takes actions on behalf of an app.
+#'    This form is most appropriate if you want to create a twitter account that
+#'    is run by a computer, rather than a human.
+#'    
+#' To use `rtweet_app()` or `rtweet_bot()` you will need to create your own 
+#' Twitter application following the instructions in `vignette("auth.Rmd")`.
+#' 
+#' See [auth_as()] to set an auth mechanism as default for this session, and
+#' [auth_save()] to save an auth mechanism so you can use it across multiple 
+#' sessions.
+#' 
+#' # Security
+#' 
+#' All of the arguments to these functions are effectively equivalent to 
+#' passwords so should ideally never be typed into the console (where they
+#' the will be recorded in `.Rhistory`) or recorded in a script (which is
+#' easy to accidentally share). Instead, the default arguments use 
+#' [askpass::askpass()] to interactively prompt you for the values.
+#' 
+#' @param api_key,api_secret Application API key and secret. These are 
+#'   generally not required for `tweet_user()` since the defaults will use
+#'   the built-in rtweet app. 
+#' @param access_token,access_secret Access token and secret.
+#' @param bearer_token App bearer token.
+#' @export
+rtweet_user <- function(api_key = NULL, api_secret = NULL) {
+  if (is.null(api_key) && is.null(api_secret)) {
+    decrypt <- function(x) {
+      rawToChar(openssl::rsa_decrypt(x[[2]], x[[1]]))
+    }
+    api_key <- decrypt(sysdat$DYKcJfBkgMnGveI)
+    api_secret <- decrypt(sysdat$MRsnZtaKXqGYHju)
+  } else {
+    stopifnot(is_string(api_key), is_string(api_secret))
+  }
+
+  app <- httr::oauth_app("rtweet", key = api_key, secret = api_secret)
+  TwitterToken1.0$new(
+    app = app,
+    endpoint = httr::oauth_endpoints("twitter"),
+    params = list(as_header = TRUE), 
+    cache_path = FALSE
+  )
+}
+
+#' @export
+#' @rdname rtweet_user
+rtweet_bot <- function(
+      api_key = askpass::askpass("Please enter your API key"),
+      api_secret = askpass::askpass("Please enter your API secret"), 
+      access_token = askpass::askpass("Please enter your access token"), 
+      access_secret = askpass::askpass("Please enter your access token")
+  ) {
+  stopifnot(is_string(api_key), is_string(api_secret))
+  stopifnot(is_string(access_token), is_string(access_secret))
```

Comment:
Provide a more informative error like missing more api_key or api_secret or...?

Reply:
Oh you're worried about the case where the user doesn't provide an info to `askpass()`? That would be better handled with a wrapper function.

Reply:
I am worried that the error won't be informative enough for some users. Not sure what happens when askpass is left empty.